odoo/modules/module.py

```diff
@@ -401,19 +403,25 @@ def is_really_module(name):
             for mname in MANIFEST_NAMES:
                 if os.path.isfile(opj(dir, name, mname)):
                     return True
-        return [
-            clean(it)
+        return {
+            clean(it): os.sep.join([ad, it])
             for it in os.listdir(dir)
             if is_really_module(it)
-        ]
+        }
 
-    plist = []
+    modules = {}
     for ad in odoo.addons.__path__:
         if not os.path.exists(ad):
             _logger.warning("addons path does not exist: %s", ad)
             continue
-        plist.extend(listdir(ad))
-    return list(set(plist))
+        modules.update(listdir(ad))
+    get_modules.__modules = frozendict(modules)
+    return get_modules.__modules
+
+def reset_modules():
+    get_modules.__modules = None
+
+reset_modules()
```

Comment:
I've a gut feeling that this can lead to thread/process synchronization problems... Like the frozenset only beeing reset in one of the threads/processes

Reply:
The __path__ update here is not threadsafe from the beginning so this pr doesn't changes much. But this operation is rare and it is unlikely to have issue with that. It is process safe tho so in worker mode there shouldn't be any issue.

Reply:
I agree that this could be change, but this is outside of the scope of this pr, i'm trying to make minimal changes here in order to avoid being stuck because changing legacy code to deeply :) But I agree that this could benefit from some cleanup :)